libs/wingsdk/src/core/plugin-manager.ts

```diff
@@ -0,0 +1,177 @@
+import { readFileSync, writeFileSync } from "fs";
+import { resolve } from "path";
+import * as vm from "vm";
+import { IConstruct } from "constructs";
+
+/**
+ * Behavioral interface of a plugin hook.
+ */
+export interface ICompilationHook {
+  /**
+   *  name of plugin can be set by plugin as export
+   *
+   * @default - absolute path to plugin file
+   */
+  pluginName: string;
+  /** preSynth phase */
+  preSynth?(app: IConstruct): void;
+  /** postSynth phase */
+  postSynth?(config: any): any;
+  /** validate phase */
+  validate?(config: any): void;
+}
+
+enum PluginPhases {
+  PRE_SYNTH = "preSynth",
+  POST_SYNTH = "postSynth",
+  VALIDATE = "validate",
+}
+
+/**
+ * Plugin manager is responsible for loading hooks from plugins and
+ * managing their executions during the various phases of the plugin lifecycle.
+ */
+export class PluginManager {
+  private hooks: ICompilationHook[];
+
+  constructor(plugins: string[]) {
+    this.hooks = [];
+
+    for (const plugin of plugins) {
+      this.add(plugin);
+    }
+  }
+
+  /**
+   * Add plugins to the plugin manager based on the plugin type.
+   *
+   * @param plugin the plugin to add
+   */
+  public add(plugin: string) {
+    // maybe we support other plugin types in the future (e.g. npm modules)
+    if (!plugin.endsWith(".js")) {
+      throw new Error(
+        `Currently only javascript files are supported as plugins. Got: ${plugin}`
+      );
+    }
+    this.addPluginFromFilePath(plugin);
```

Comment:
Why do we need both `add()` and `addPluginFromFilePath()`?

Reply:
The thought here was that maybe plugins are not always absolute paths to some JS file. Maybe we have some wing standard plugins that users can choose to toggle like "tag-resources" or "s3-backend" etc.. so in this case the add() method would control how we add plugins, (as absolute paths or something else). 

Though right now its not useful.